nvs: also support sector_size of 64KB

mjchen0 opened the pull request:
Allows NVS to work with flash device configured to use only 64KB block erase.

mjchen0 committed:
nvs: support sector_size of 64KB

Allows NVS to work with flash device configured to
use only 64KB block erase. Due to how addresses
are encoded internally in NVS, 64KB is the maximum
sector size. Add a test for this during mount.

Add a native_sim unit test case for 64kb erase block size

Signed-off-by: Mike J. Chen <mjchen@google.com>

Laczen commented:
Hi @mjchen0, the limitation of NVS to smaller sector sizes is an internal property. The proposed change does not enable the use of larger sectors.

Laczen commented:
> > Hi @mjchen0, the limitation of NVS to smaller sector sizes is an internal property. The proposed change does not enable the use of larger sectors.
> 
> Sorry, my title wasn't correct (I've changed it). I wasn't trying to support arbitrarily large sectors. I was just trying to make NVS work with FLASH that has only 64KB erase units. NVS seems to have a max sector size of 64KB. It's not described in the documentation of nvs.rts, but appears due to the representation of address as described in nvs_priv.h:
> 
> ```
> /*                                                                                                                                 
>  * MASKS AND SHIFT FOR ADDRESSES                                                                                                   
>  * an address in nvs is an uint32_t where:                                                                                         
>  *   high 2 bytes represent the sector number                                                                                      
>  *   low 2 bytes represent the offset in a sector                                                                                  
>  */                                                                                                                                
> #define ADDR_SECT_MASK 0xFFFF0000                                                                                                  
> #define ADDR_SECT_SHIFT 16                                                                                                         
> #define ADDR_OFFS_MASK 0x0000FFFF 
> ```
> 
> With 16-bits for offset, a 64KB sector size is possible (offset 0 to 0xffff) except that the sector_size field itself is uint16_t so 64KB couldn't be stored as the size.

Ok, makes sense. Could you extend the PR with a check that sector_size is <= 64kB ? Also change the title to reflect "<=64kB".

mjchen0 commented:
I added the test in nvs_mount() that the sector_size isn't greater than 64KB.

I also added a new test case for native_sim with 64KB erase blocks, and a test that the nvs_mount() returns the expected error for a bad sector_size value.

Laczen reviewed (dismissed):
Thanks

Laczen reviewed (approved):
Thanks

Laczen commented:
@nashif, @de-nordic, please add a review.
